src/Raven.Server/Documents/Revisions/RevisionsStorage.cs

```diff
@@ -2477,8 +2477,72 @@ public IEnumerable<Document> GetRevisionsBinEntries(DocumentsOperationContext co
             }
         }
 
+        public List<(string Id, long Etag)> GetDeletedRevisionsIds(DocumentsOperationContext context, DateTime before,
+            long batchSize, ref long lastEtag, CancellationToken token)
+        {
+            var ids = new List<(string Id, long Etag)>();
+            var count = 0L;
+            var table = new Table(RevisionsSchema, context.Transaction.InnerTransaction);
+            using var _ = GetEtagAsSlice(context, lastEtag, out var startSlice);
+
+            var enumerator = new TransactionForgetAboutStorageIdEnumerator(
+                table.SeekForwardFrom(RevisionsSchema.Indexes[DeleteRevisionEtagSlice], startSlice, 0).GetEnumerator(), context);
+
+            foreach (var result in enumerator/*table.SeekForwardFrom(RevisionsSchema.Indexes[DeleteRevisionEtagSlice], startSlice, 0)*/)
+            {
+                ref var reader = ref result.Result.Reader;
+                var deleteRevision = TableValueToRevision(context, ref reader, DocumentFields.Id);
+                var deletedEtag = TableValueToEtag((int)RevisionsTable.DeletedEtag, ref reader);
+                lastEtag = deletedEtag;
+                if (deleteRevision.LastModified >= before)
+                    break;
+
+                ids.Add((deleteRevision.Id, deletedEtag));
+
+                if (count++ > batchSize || token.IsCancellationRequested)
+                    break;
+            }
+
+            return ids;
+        }
+
+        public static long ReadLastRevisionsBinCleanerLastEtag(Transaction tx)
+        {
+            if (tx == null)
+                throw new InvalidOperationException("No active transaction found in the context, and at least read transaction is needed");
+            var tree = tx.ReadTree(DocumentsStorage.GlobalTreeSlice);
+            if (tree == null)
```

Comment:
can this ever happened that the parent tree doesn't exists?

Reply:
You are right, it can't be null
entered here for the first time in debug and the tree wasn't null.